civis/ml/_model.py

```diff
@@ -0,0 +1,443 @@
+        self.metadata = metadata
+        self.estimator = estimator
+        super().__init__(msg)
+
+
+def _check_is_fit(method):
+    """Makes sure that the ModelPipeline's been trained"""
+    @wraps(method)
+    def wrapper(*args, **kwargs):
+        self = args[0]
+        if not self.train_result_:
+            raise ValueError("This model hasn't been trained yet.")
+        return method(*args, **kwargs)
+    return wrapper
+
+
+def _block_and_handle_missing(method):
+    """For ModelFuture file-retrieving property methods.
+    Block until completion and attempt to retrieve result.
+    Raise exception only if the result isn't found.
+    """
+    @wraps(method)
+    def wrapper(self):
+        futures.wait((self,))  # Block until done
+        try:
+            return method(self)
+        except FileNotFoundError:
+            # We get here if the modeling job failed to produce
+            # any output and we don't have metadata.
+            if self.exception():
+                raise self.exception() from None
+            else:
+                raise
+    return wrapper
+
+
+def _stash_local_data(X, client=None):
+    """Store data in a temporary Civis File and return the file ID
+    """
+    civis_fname = 'modelpipeline_data.csv'
+    if HAS_PANDAS and isinstance(X, pd.DataFrame):
+        buf = io.BytesIO()
+        txt = io.TextIOWrapper(buf, encoding='utf-8')
+        X.to_csv(txt, encoding='utf-8', index=False)
+        txt.flush()
+        buf.seek(0)
+        file_id = cio.file_to_civis(buf, name=civis_fname, client=client)
+    else:
+        with open(X) as _fin:
+            file_id = cio.file_to_civis(_fin, name=civis_fname, client=client)
+
+    return file_id
+
+
+def _decode_train_run(train_job_id, train_run_id, client):
+    """Determine correct run ID for use for a given training job ID"""
+    try:
+        return int(train_run_id)
+    except ValueError:
+        container = client.scripts.get_containers(train_job_id)
+        if train_run_id == 'active':
+            train_run_id = container.arguments.get('ACTIVE_BUILD', find_one(
+                container.params, name='ACTIVE_BUILD'))['default']
+
+        if train_run_id == 'latest':
+            return container.last_run.id
+
+        try:
+            return int(train_run_id)
+        except Exception as exc:
+            raise ValueError('Please provide valid train_run_id! Needs to be '
+                             'integer corresponding to a training run ID '
+                             'or one of "active" or "latest".') from exc
+
+
+def _retrieve_file(fname, job_id, run_id, local_dir, client=None):
+    """Download a Civis file using a reference on a previous run
+
+    Parameters
+    ----------
+    fname: str
+        Name of the Civis file you wish to retrieve
+    job_id, run_id: int
+        Job ID and run ID of the run holding a reference to this file
+    local_dir: str
+        Download the file `fname` from Civis Platform to this directory
```

Comment:
I removed the parameters section of this docstring, since it's a private function.